U19 notice page draft (#20128)
ethereum-optimism/optimism: docs/public-docs/notices/upgrade-19.mdx

---
title: Upgrade 19 - L2 Contract Manager, Fusaka on L2, and Kona Respected Game Type
description: Learn how to prepare for Upgrade 19.
---

Upgrade 19 is a proposed network upgrade for OP Stack chains. It **introduces the L2 Contract Manager (L2CM)** enabling governance-approved L2 smart contract upgrades through a new onchain mechanism, **promotes `CANNON_KONA` to the respected game type** making the Rust-based `kona-client` the primary fault proof program used for withdrawals, **activates Fusaka EIPs on L2** including the EIP-7825 transaction gas limit, and **adds a preemptive Glamsterdam Defense change** limiting BN256 pairing precompile input sizes.

<Info>

>   Upgrade 19 includes hard fork changes (Fusaka on L2 and Glamsterdam Defense). Node operators must upgrade before the activation timestamp. Activation timestamps for each network will be communicated once finalized.

Reviewer: Let's change the name to Osaka on L2. I'll change the name of the design doc as well.

Edit: [updated the naming](https://github.com/ethereum-optimism/design-docs/blob/main/protocol/osaka-on-l2.md) so you'll want to change the link as well.

Upgrade 19 introduces the following changes:

* **L2 Contract Manager (L2CM):** introduces an onchain mechanism for upgrading L2 predeploy smart contracts as part of a network upgrade. L2CM enables governance to approve and execute L2 predeploy upgrades in a structured, auditable way — reducing the need for a hard fork for each individual contract change. This is a prerequisite for interop and future protocol upgrades that need to modify L2 contracts. See the [L2CM design doc](https://github.com/ethereum-optimism/design-docs/blob/main/protocol/l2-contract-upgrades.md) and [FMA](https://github.com/ethereum-optimism/design-docs/blob/main/security/fma-l2cm.md) for details.

* **Fusaka on L2:** activates Fusaka EIPs on L2, including [EIP-7825](https://eips.ethereum.org/EIPS/eip-7825) which introduces a per-transaction gas limit. Deposit transactions are **exempt** from the EIP-7825 gas limit because they are already capped at 20M gas per L1 block — enforcing the limit on deposits could cause permanent loss of deposited ETH. See the [Fusaka on L2 design doc](https://github.com/ethereum-optimism/design-docs/blob/main/protocol/fusaka-on-l2.md) for details.

> * **Glamsterdam Defense (BN256 pairing input size limitation):** as a preemptive measure ahead of the Glamsterdam L1 hard fork, Upgrade 19 includes a hard fork change that limits the input size for BN256 pairing precompile calls. The full Glamsterdam Defense smart contract changes (absolute pre-state upgrade) are deferred and will be slotted in once final L1 gas pricing values are confirmed.

Reviewer: @pauldowman have we confirmed the numbers that bn256 is the only precompile at risk? Last I checked are others are are also pretty close but maybe I misunderstood.

Reviewer 2: It's the only one with proposed changes that is also variable input, which is the thing that matters (since the fixed-length input ones don't create risk of hitting the 16Mgas L1 tx limit. Details [here](https://www.notion.so/oplabs/EXT-Precompile-Oracle-Gas-Cost-Risks-in-OP-Stack-Fault-Proofs-318f153ee162812c8675c10482ffb6c7?source=copy_link#b3c63ee5a5f544d5855a6913e6213edf).

Reviewer: Right, just want to triple check that we've confirmed the numbers in the doc are accurate... I can do that if you like (probably would need to be reviewed by a proofs team member anyway) but iirc those are claude generated.

Reviewer: [(source)](https://oplabs-pbc.slack.com/archives/C07LJL1E0FJ/p1772582000970389?thread_ts=1772560363.090169&cid=C07LJL1E0FJ)

* **New Kona absolute prestate:** ships an updated `cannon64-kona` absolute prestate hash embedding the latest chain configurations.

* **OPCMv2:** Upgrade 19 is the first upgrade to use OPCMv2, the redesigned OP Contracts Manager.

<Info>
  This upgrade uses OPCMv2, the new default OP Contracts Manager. Chain operators calling `OPCM.upgrade()` directly or using `op-deployer` should ensure they are targeting the correct OPCMv2 contract version.
</Info>

## Breaking Changes

### Fusaka on L2: EIP-7825 transaction gas limit

After Upgrade 19, L2 transactions are subject to the [EIP-7825](https://eips.ethereum.org/EIPS/eip-7825) per-transaction gas limit. Transactions requesting gas above this limit will be invalid.

* **Deposit transactions are exempt** from the EIP-7825 limit. Deposits are already capped at 20M gas total per L1 block, and rejecting deposits on L2 that were accepted on L1 could cause permanent ETH loss. Deposits will continue to land on L2 even if their gas limit exceeds the EIP-7825 threshold.

> * **For app developers:** if your application submits L2 transactions with very high gas limits (above the EIP-7825 limit), those transactions will no longer be valid after this upgrade. In practice, very few transactions are expected to exceed this limit.

Reviewer: Let's also mention the changes in gas amounts for MODEXP + p256verify and the input size limit for MODEXP. Those are breaking changes that app developers should be aware of as well.

(these notices should be present in the app developers section below as well)

* **MODEXP and P256VERIFY gas changes (Osaka on L2):** the gas costs for `MODEXP` (`0x05`) and `P256VERIFY` (`0x100`) precompiles are updated. A new input size limit is also introduced for `MODEXP`. If your contracts call these precompiles, update your gas estimates and verify inputs are within the new limits.
* **BN256 pairing input size limit (Glamsterdam Defense):** the maximum input for `ecPairing` (`0x08`) is reduced from 427 to 390 pairs (74,880 bytes). If your contracts call the BN256 pairing precompile with more than 390 pairs, those calls will fail. Standard ZK proof verification (e.g., Groth16) is not affected.
* **L2 predeploy upgrades:** if your application interacts directly with predeploy contracts (e.g., `L2ToL1MessagePasser`, `L2CrossDomainMessenger`, `L2StandardBridge`), review the changelog for new functions or behavior changes.
* **Withdrawal proving:** the respected game type changes from `CANNON` (0) to `CANNON_KONA` (8). Standard SDK usage handles this automatically.

## Execute the L1 Contract Upgrade

> Once your infrastructure is ready, execute the upgrade transaction. This should be done by making a delegatecall to the `upgrade()` function of the OP Contracts Manager (OPCMv2) at the address listed in [the registry](https://github.com/ethereum-optimism/superchain-registry/blob/main/validation/standard/standard-versions-mainnet.toml).

Reviewer 3: ```suggestion
Once your infrastructure is ready, execute the upgrade transaction. This should be done by making a delegatecall to the `upgrade()` function of the OP Contracts Manager (OPCMv2) at the address that will be listed in [the registry](https://github.com/ethereum-optimism/superchain-registry/blob/main/validation/standard/standard-versions-mainnet.toml).
```

---
title: Upgrade 19 - L2 Contract Manager, Osaka on L2, and Kona Respected Game Type
description: Learn how to prepare for Upgrade 19.
---

> Upgrade 19 is a proposed network upgrade for OP Stack chains. It **introduces the L2 Contract Manager (L2CM)** enabling governance-approved L2 smart contract upgrades through a new onchain mechanism, **promotes `CANNON_KONA` to the respected game type** making the Rust-based `kona-client` the primary fault proof program used for withdrawals, **activates Osaka EIPs on L2** including the EIP-7825 transaction gas limit, and **adds a preemptive Glamsterdam Defense change** reducing the BN256 pairing precompile maximum input size.

Reviewer 4: ```suggestion
Upgrade 19 is a proposed network upgrade for OP Stack chains. It **introduces the L2 Contract Manager (L2CM)** enabling governance-approved L2 smart contract upgrades through a new consensus layer mechanism, **promotes `CANNON_KONA` to the respected game type** making the Rust-based `kona-client` the primary fault proof program used for withdrawals, **activates Osaka EIPs on L2** including the EIP-7825 transaction gas limit, and **adds a preemptive Glamsterdam Defense change** reducing the BN256 pairing precompile maximum input size.
```

<Info>

>   Upgrade 19 includes hard fork changes (Osaka on L2 and Glamsterdam Defense). Node operators must upgrade before the activation timestamp. Activation timestamps for each network will be communicated once finalized.

Reviewer 4: ```suggestion
  Upgrade 19 includes hard fork changes (Osaka on L2, Glamsterdam Defense and hard fork activated L2 Contract upgrades). Node operators must upgrade before the activation timestamp. Activation timestamps for each network will be communicated once finalized.
```

</Info>

## What's included in Upgrade 19

Upgrade 19 introduces the following changes:

> * **L2 Contract Manager (L2CM):** introduces an onchain mechanism for upgrading L2 predeploy smart contracts as part of a network upgrade. L2CM enables governance to approve and execute L2 predeploy upgrades in a structured, auditable way — reducing the need for a hard fork for each individual contract change. This is a prerequisite for interop and future protocol upgrades that need to modify L2 contracts. See the [L2CM design doc](https://github.com/ethereum-optimism/design-docs/blob/main/protocol/l2-contract-upgrades.md) and [FMA](https://github.com/ethereum-optimism/design-docs/blob/main/security/fma-l2cm.md) for details.

Reviewer 4: ```suggestion
* **L2 Contract Manager (L2CM):** introduces a mechanism for upgrading L2 smart contracts (predeploys) as part of a network upgrade. L2CM enables L2 predeploy upgrades in a structured, auditable way — reducing the need for multisig signing for each individual contract change. This is a prerequisite for interop and future protocol upgrades that need to modify L2 contracts. See the [L2CM design doc](https://github.com/ethereum-optimism/design-docs/blob/main/protocol/l2-contract-upgrades.md) and [FMA](https://github.com/ethereum-optimism/design-docs/blob/main/security/fma-l2cm.md) for details.
```

* **`CANNON_KONA` set to respected game type:** changes the respected game type from `CANNON` (0) to `CANNON_KONA` (8), making the Rust-based `kona-client` (`kona-node` + `op-reth`) the primary fault proof program used for withdrawals. The `CANNON` (0) game type is retained as a fallback alongside `op-program` until the full transition to `op-reth` is complete. This builds on the `CANNON_KONA` game type support introduced in [Upgrade 18](/notices/archive/upgrade-18).

* **New Kona absolute prestate:** ships an updated `cannon64-kona` absolute prestate hash embedding the latest chain configurations.

* **OPCMv2:** Upgrade 19 is the first upgrade to use OPCMv2, the redesigned OP Contracts Manager.

<Info>

>   This upgrade uses OPCMv2, the new default OP Contracts Manager. Chain operators calling `OPCM.upgrade()` directly or using `op-deployer` should ensure they are targeting the correct OPCMv2 contract version.

Reviewer 3: I'd add the following to this:

OPCMv2 has a new interface, please take a look at the source code for more information (link the OPCMv2 source code)

* **For chain operators running permissionless fault proofs:** `op-challenger` must be configured to support `cannon-kona` trace types **prior** to the upgrade. Failure to update `op-challenger` may result in uncontested invalid games.

### L2CM: new L2 predeploy upgrade mechanism

L2CM introduces a new mechanism for upgrading L2 predeploy contracts. App developers building on OP Stack chains should be aware that predeploy contracts may now be upgraded through L2CM transactions. If your application interacts directly with predeploy contracts, review the changelog for each upgrade to understand new capabilities.

## For chain operators

Update the following components:

| Component | Version | Notes |
| ---------------- | ------- | ----- |
| `op-node`        | `TBD`   | Must be updated **prior** to the activation timestamp. Required for Osaka on L2 and Glamsterdam Defense hard fork changes. |
| `op-reth`        | `TBD`   | Must be updated **prior** to the activation timestamp. Required for EIP-7825 gas limit enforcement, MODEXP/P256VERIFY gas changes, and BN256 pairing input size reduction. |

> | `op-geth`        | `TBD`   | Must be updated **prior** to the activation timestamp if still running `op-geth`. Required for EIP-7825 gas limit enforcement, MODEXP/P256VERIFY gas changes, and BN256 pairing input size reduction. |

Reviewer 3: let's kill op-geth version

> | `op-geth`        | `TBD`   | Must be updated **prior** to the activation timestamp if still running `op-geth`. Required for EIP-7825 gas limit enforcement, MODEXP/P256VERIFY gas changes, and BN256 pairing input size reduction. |

Reviewer 3: ```suggestion
```
